src/Init/Lean/Elab/Quotation.lean

```diff
@@ -6,65 +6,111 @@ Authors: Sebastian Ullrich
 prelude
 import Init.Lean.Syntax
 import Init.Lean.Elab.ResolveName
+import Init.Lean.Elab.Term
 import Init.Lean.Parser -- TODO: remove after removing old elaborator
 
 namespace Lean
-namespace Elab
 
--- TODO: replace with quotations where possible
-private def const (n : Name) : Syntax :=
-Syntax.node `Lean.Parser.Term.id #[Syntax.ident none n.toString.toSubstring n [(n, [])]]
+/-- A monad that supports syntax quotations. -/
+class MonadQuotation (m : Type → Type) :=
+-- Get the fresh scope of the current macro invocation
+(getCurrMacroScope : m Nat)
+-- Execute action in a new macro invocation context
+(withFreshMacroScope {α : Type} : m α → m α)
+export MonadQuotation
 
-private def app (fn arg : Syntax) : Syntax :=
-Syntax.node `Lean.Parser.Term.app #[fn, arg]
+/-- Simplistic MonadQuotation that does not guarantee globally fresh names. It is only safe
+    if the syntax quotations do not introduce bindings around antiquotations, and
+    if references to globals are prefixed with `_root_.`. -/
+abbrev Unhygienic := ReaderT (List Nat) $ StateM Nat
+namespace Unhygienic
+instance MonadQuotation : MonadQuotation Unhygienic := {
+  getCurrMacroScope := do
+    stack ← read;
+    pure $ stack.head!,
+  withFreshMacroScope := fun α x => do
+    fresh ← modifyGet (fun n => (n, n + 1));
+    adaptReader (fun stack => fresh::stack) x
+}
+protected def run {α : Type} (x : Unhygienic α) : α := run x [0] 1
+end Unhygienic
```

Comment:
@leodemoura Do this class and its sample implementation make sense to you? If so, we would eventually need the same setup in the elaboration monads. Note that I haven't implemented the mangling scheme yet, but I don't think it should be part of this class.

Reply:
`withFreshMacroScope` would be used at `elabTerm` and other recursion points. But it can also be useful for creating local scopes in a single complex macro.

Reply:
I like the idea of having a `MonadQuotation` class. The two methods look reasonable to me, but I need to see them in "action" to fully appreciate the design. AFAICT, they are not being used anywhere right now, correct?

Reply:
BTW, syntax quotations are helpful for implementing the elaborator even if we don't have `match_syntax`. For example, it is easy to decode programmatically the Lean4 `lambda` syntax without `match_syntax`, but it is quite annoying to construct the right-hand side of the rule `fun <pat> => <term>` --> `fun x => match x with | <pat> => term` without syntax quotations.

Reply:
> BTW, syntax quotations are going to be more helpful than I expected

Great to hear :) . Dealing with syntax objects manually really is a royal pain...